uberfire-extensions/uberfire-widgets/uberfire-widgets-core/uberfire-widgets-core-client/src/main/java/org/uberfire/ext/widgets/core/client/tree/FSTreeItem.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2017 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.uberfire.ext.widgets.core.client.tree;
+
+import java.util.function.Supplier;
+
+import com.google.gwt.user.client.ui.FlowPanel;
+import org.gwtbootstrap3.client.ui.Icon;
+import org.gwtbootstrap3.client.ui.constants.IconType;
+
+import static org.uberfire.commons.validation.PortablePreconditions.checkNotNull;
+
+public class FSTreeItem extends TreeItem {
```

Comment:
Once adding the generics for `TreeItem` class (see comments on `TreeItem` below), you can here change this by:
`public class FSTreeItem extends TreeItem<FSTreeItem>`
This way, for example on the `TreeNavigator` class in `kie-wb-common`, that uses this, there is no need to do casts for methods like `getChildren()` or` getChild(index)`.